core/src/test/java/kafka/server/share/DelayedShareFetchTest.java

```diff
@@ -261,10 +297,99 @@ public void testToCompleteAnAlreadyCompletedFuture() {
         Mockito.verify(delayedShareFetch, times(1)).acquirablePartitions();
     }
 
+    @Test
+    public void testForceCompleteTriggersDelayedActionsQueue() {
+        String groupId = "grp";
+        Uuid topicId = Uuid.randomUuid();
+        ReplicaManager replicaManager = mock(ReplicaManager.class);
+        TopicIdPartition tp0 = new TopicIdPartition(topicId, new TopicPartition("foo", 0));
+        TopicIdPartition tp1 = new TopicIdPartition(topicId, new TopicPartition("foo", 1));
+        TopicIdPartition tp2 = new TopicIdPartition(topicId, new TopicPartition("foo", 2));
+        Map<TopicIdPartition, Integer> partitionMaxBytes1 = new HashMap<>();
+        partitionMaxBytes1.put(tp0, PARTITION_MAX_BYTES);
+        partitionMaxBytes1.put(tp1, PARTITION_MAX_BYTES);
+
+        SharePartition sp0 = mock(SharePartition.class);
+        SharePartition sp1 = mock(SharePartition.class);
+        SharePartition sp2 = mock(SharePartition.class);
+        // No share partition is available for acquiring initially.
+        when(sp0.maybeAcquireFetchLock()).thenReturn(false);
+        when(sp1.maybeAcquireFetchLock()).thenReturn(false);
+        when(sp2.maybeAcquireFetchLock()).thenReturn(false);
+
+        Map<SharePartitionKey, SharePartition> partitionCacheMap = new ConcurrentHashMap<>();
+        partitionCacheMap.put(new SharePartitionKey(groupId, tp0), sp0);
+        partitionCacheMap.put(new SharePartitionKey(groupId, tp1), sp1);
+        partitionCacheMap.put(new SharePartitionKey(groupId, tp2), sp2);
+
+        ShareFetchData shareFetchData1 = new ShareFetchData(
+            new FetchParams(ApiKeys.SHARE_FETCH.latestVersion(), FetchRequest.ORDINARY_CONSUMER_ID, -1, MAX_WAIT_MS,
+                1, 1024 * 1024, FetchIsolation.HIGH_WATERMARK, Optional.empty()), groupId, Uuid.randomUuid().toString(),
+            new CompletableFuture<>(), partitionMaxBytes1);
+
+        DelayedOperationPurgatory<DelayedShareFetch> delayedShareFetchPurgatory = new DelayedOperationPurgatory<>(
+            "TestShareFetch", mockTimer, replicaManager.localBrokerId(),
+            DELAYED_SHARE_FETCH_PURGATORY_PURGE_INTERVAL, true, true);
+
+        Set<Object> delayedShareFetchWatchKeys = new HashSet<>();
+        partitionMaxBytes1.keySet().forEach(topicIdPartition -> delayedShareFetchWatchKeys.add(new DelayedShareFetchKey(groupId, topicIdPartition)));
+
+        DelayedShareFetch delayedShareFetch1 = DelayedShareFetchTest.DelayedShareFetchBuilder.builder()
+            .withShareFetchData(shareFetchData1)
+            .withReplicaManager(replicaManager)
+            .withPartitionCacheMap(partitionCacheMap)
+            .build();
+
+        // We add a delayed share fetch entry to the purgatory which will be waiting for completion since none of the
```

Comment:
There are only 2 partitions in delayedShareFetch1. So we should use neither instead none?